client/components/web-preview/content.jsx

```diff
@@ -0,0 +1,309 @@
+import classNames from 'classnames';
+import debugModule from 'debug';
+import noop from 'lodash/noop';
+import page from 'page';
+import shallowCompare from 'react-addons-shallow-compare';
+import { v4 as uuid } from 'uuid';
+import addQueryArgs from 'lib/route/add-query-args';
+
+/**
+ * Internal dependencies
+ */
+import Toolbar from './toolbar';
+import touchDetect from 'lib/touch-detect';
+import { isMobile } from 'lib/viewport';
+import { localize } from 'i18n-calypso';
+import Spinner from 'components/spinner';
+import SeoPreviewPane from 'components/seo-preview-pane';
+import { recordTracksEvent } from 'state/analytics/actions';
+
+const debug = debugModule( 'calypso:web-preview' );
+
+export class WebPreviewContent extends Component {
+	constructor( props ) {
+		super( props );
+
+		this.previewId = uuid();
+
+		this._hasTouch = false;
+		this._isMobile = false;
+
+		this.state = {
+			iframeUrl: null,
+			device: props.defaultViewportDevice || 'computer',
+			loaded: false
+		};
+
+		this.setIframeInstance = ref => this.iframe = ref;
+	}
+
+	componentWillMount() {
+		// Cache touch and mobile detection for the entire lifecycle of the component
+		this._hasTouch = touchDetect.hasTouch();
+		this._isMobile = isMobile();
+	}
+
+	componentDidMount() {
+		window.addEventListener( 'message', this.handleMessage );
+		if ( this.props.previewUrl ) {
+			this.setIframeUrl( this.props.previewUrl );
+		}
+		if ( this.props.previewMarkup ) {
+			this.setIframeMarkup( this.props.previewMarkup );
+		}
+
+		if ( typeof this.props.onDeviceUpdate === 'function' ) {
+			this.props.onDeviceUpdate( this.state.device );
+		}
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'message', this.handleMessage );
+	}
+
+	shouldComponentUpdate( nextProps, nextState ) {
+		return shallowCompare( this, nextProps, nextState );
+	}
+
+	componentDidUpdate( prevProps ) {
+		const { previewUrl } = this.props;
+
+		this.setIframeUrl( previewUrl );
+
+		// If the previewMarkup changes, re-render the iframe contents
+		if ( this.props.previewMarkup && this.props.previewMarkup !== prevProps.previewMarkup ) {
+			this.setIframeMarkup( this.props.previewMarkup );
+		}
+		// If the previewMarkup is erased, remove the iframe contents
+		if ( ! this.props.previewMarkup && prevProps.previewMarkup ) {
+			debug( 'removing iframe contents' );
+			this.setIframeMarkup( '' );
+		}
+		// Focus preview when showing modal
+		if ( this.props.showPreview && ! prevProps.showPreview && this.state.loaded ) {
+			this.focusIfNeeded();
+		}
+	}
+
+	handleMessage = ( e ) => {
+		let data;
+		try {
+			data = JSON.parse( e.data );
+		} catch ( err ) {
+			return;
+		}
+
+		if ( ! data || data.channel !== 'preview-' + this.previewId ) {
+			return;
+		}
+		debug( 'message from iframe', data );
+
+		switch ( data.type ) {
+			case 'link':
+				page( data.payload.replace( /^https:\/\/wordpress\.com\//i, '/' ) );
+				if ( typeof this.props.onClose === 'function' ) {
```

Comment:
Since there are quite a few of these tests, I suggest `{ isFunction } from 'lodash'`